### PR TITLE
Add `dq.smesolve()`

### PR DESCRIPTION
**Pull request by pierreguilmin**
Closes DYN-164.

First version April 11, 2024, updated September 2, 2024. Changed target branch to `dev-smesolve`.

This is quite a big one, I advise reading commit by commit. I did my best to split the logic between saving, solver logic and API. This solver has not reached it's final version, consider this is a first working POC on which we can iterate.

Screenshot 2024-04-11 at 22 59 48

## Example API usage
```python
H = dq.zero(2)
psi0 = dq.unit(dq.basis(2, 0) + dq.basis(2, 1))
jump_ops = [1.0 * dq.sigmaz()]
etas = [1.0]
tsave = np.linspace(0, 1.0, 101)
key = jax.random.PRNGKey(42)
ntrajs = 6
keys = jax.random.split(key, ntrajs)

result = dq.smesolve(H, jump_ops, etas, psi0, tsave, keys)

print(result.states.shape)  # (ntrajs, ntsave, n, n)
print(result.measurements.shape)  # (ntrajs, nLm, ntsave-1)
```

## Example 1: monitored qubit
First example to convince the SME-practionner that the solver is correct. Qubit starting from $\ket+$ with $H=0$ and monitored $L=\sigma_z$ (we expect diffusive projection onto one of $\sigma_z$ eigenvalues).

qubit-trajectories

<details>
  <summary>Code</summary>

  ```python
  # imports
  import jax
  import jax.numpy as jnp
  import numpy as np
  
  import dynamiqs as dq
  from dynamiqs.plots.utils import mplstyle
  
  mplstyle()
  
  # simulation
  H = dq.zero(2)
  psi0 = dq.unit(dq.basis(2, 0) + dq.basis(2, 1))
  jump_ops = [1.0 * dq.sigmaz()]
  etas = [1.0]
  tsave = np.linspace(0, 1.0, 101)
  key = jax.random.PRNGKey(42)
  ntrajs = 6
  keys = jax.random.split(key, ntrajs)
  
  result = dq.smesolve(H, jump_ops, etas, psi0, tsave, keys)
  
  print(result.states.shape)  # (ntrajs, ntsave, n, n)
  print(result.measurements.shape)  # (ntrajs, nLm, ntsave-1)
  
  # plot
  fig, axs = dq.gridplot(3, 3, w=6, h=3, sharex=True)
  ax0, ax1, ax2 = list(axs)
  
  for states in result.states:
      ax0.plot(tsave, dq.expect(dq.sigmaz(), states).real, lw=1.5)
      ax0.axhline(-1, ls='--', lw=1.0, color='gray')
      ax0.axhline(1, ls='--', lw=1.0, color='gray')
      ax0.set(
          title=r'Time evolution of $\sigma_z$ expectation value',
          xlabel=r'$t$',
          ylabel=r'$\mathrm{Tr}[\sigma_z\rho_t]$',
          ylim=(-1.1, 1.1),
      )
  
  for It in result.measurements[:, 0]:
      integrated_signal = jnp.cumsum(It)
      ax1.axhline(0, ls='--', lw=1.0, color='gray')
      ax1.plot(tsave[1:], integrated_signal, lw=1.5)
      ax1.set(
          title=r'Time integrated signal',
          xlabel=r'$t$',
          ylabel=r'$\int_0^t \mathrm{d}Y_u$',
          ylim=(-4.5, 4.5),
      )
  
      ax2.plot(tsave[:-1], It, lw=1.5)
      delta_t = tsave[1] - tsave[0]
      ax2.set(
          title=rf'Instantaneous signal (binned on $\Delta t={delta_t}$)',
          xlabel=r'$t$',
          ylabel=r'$I_t$',
          ylim=(-0.4, 0.4),
      )
  
  fig.savefig('qubit-sme.png', dpi=300, bbox_inches='tight')
  ```

</details>

## Example 2: monitored cavity

Cavity starting from $\ket\alpha$ with $H=\Delta a^\dagger a$ and monitored $L_X = \sqrt\kappa a$ and $L_P = \sqrt\kappa (-ia)$ (we expect no measurement backaction and average signal trajectories following the state expectation values)
cavity-sme

<details>
  <summary>Code</summary>

```python
# imports
import jax
import jax.numpy as jnp
import numpy as np

import dynamiqs as dq
from dynamiqs.plots.utils import mplstyle

mplstyle()

# simulation
n = 16
a = dq.destroy(n)
H = 10 * dq.dag(a) @ a
psi0 = dq.coherent(n, 2.0)
jump_ops = [a, -1j * a]
etas = [1.0, 1.0]
tsave = np.linspace(0, 1.0, 101)
key = jax.random.PRNGKey(42)
ntrajs = 100
keys = jax.random.split(key, ntrajs)

result = dq.smesolve(H, jump_ops, etas, psi0, tsave, keys)

# plot
fig, axs = dq.gridplot(6, 3, w=6, h=3, sharex=True)
ax0, ax1, ax2, ax3, ax4, ax5 = list(axs)
delta_t = tsave[1] - tsave[0]

for states in result.states:
    ax0.plot(tsave, dq.expect(dq.position(n), states).real, lw=1.5)
    ax0.set(
        title=r'Time evolution of $X$ expectation value',
        xlabel=r'$t$',
        ylabel=r'$\mathrm{Tr}[X\rho_t]$',
        ylim=(-2.5, 2.5),
    )

    ax1.plot(tsave, dq.expect(dq.momentum(n), states).real, lw=1.5)
    ax1.set(
        title=r'Time evolution of $P$ expectation value',
        xlabel=r'$t$',
        ylabel=r'$\mathrm{Tr}[P\rho_t]$',
        ylim=(-2.5, 2.5),
    )

ItXs = result.measurements[:, 0] / delta_t / 2
ax2.plot(tsave[1:], jnp.mean(ItXs, axis=0))
ax2.set(
    title=r'$X$ quadrature signal averaged over all trajectories',
    xlabel=r'$t$',
    ylabel=r'$\mathbb{E}[I_t^X]$',
    ylim=(-2.5, 2.5),
)

ItPs = result.measurements[:, 1] / delta_t / 2
ax3.plot(tsave[1:], jnp.mean(ItPs, axis=0))
ax3.set(
    title=r'$P$ quadrature signal averaged over all trajectories',
    xlabel=r'$t$',
    ylabel=r'$\mathbb{E}[I_t^P]$',
    ylim=(-2.5, 2.5),
)

for It in ItXs[:3]:
    ax4.plot(tsave[:-1], It, lw=1.5)
    ax4.set(
        title=rf'Instantaneous $X$ quadrature signal for 3 trajectories (binned on $\Delta t={delta_t}$)',
        xlabel=r'$t$',
        ylabel=r'$I_t^X$',
        ylim=(-15, 15),
    )

for It in ItPs[:3]:
    ax5.plot(tsave[:-1], It, lw=1.5)
    ax5.set(
        title=rf'Instantaneous $P$ quadrature signal for 3 trajectories (binned on $\Delta t={delta_t}$)',
        xlabel=r'$t$',
        ylabel=r'$I_t^P$',
        ylim=(-15, 15),
    )

fig.savefig('cavity-sme.png', dpi=300, bbox_inches='tight')
```
</details>

## Miscellaneous
I manually tested `H` and `rho0` batching, that `exp_ops` keep working, and that `tmeas` does its job.

What's missing:
- (easy to solve) the same key is used by each batch instance
- (easy to solve) result.wiener is not correctly batched
- (tedious but easy) write tests
- (meeeh) the outer smesolve function is not JIT-compatible due to the split between measured and dissipative operators
- (meeeeh) performance is not very good

### CODE REVIEW

**Comment by pierreguilmin** [comment]
Follow-up issue: https://linear.app/dynamiqs/issue/DYN-196/improve-dqsmesolve.

File: dynamiqs/result.py [hunk 1]
```diff
@@ -120,3 +120,48 @@ class MEResult(Result):
...
+class SMESaved(Saved):
+    Isave: Array
+    wiener: PyTree
+
+
+class SMEResult(Result):
+    """Result of the SME integration.
+
+    Attributes:
+        states _(array of shape (nH?, nrho0?, ntrajs, ntsave, n, n))_: Saved states.
```
Comment (gautierronan) [inline dynamiqs/result.py:134]: It's the same in `MEResult`, but I guess we could also batch over jump operators. In any case, we'll have to properly document this at some point: it's very much related to the discussion on TimeArrays.

I'm thinking a notation like `(nH..., nLs..., nrho0..., ntrajs, ntsave, n, n)` could be more explicit than what we currently have.

File: dynamiqs/smesolve/smediffrax.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+# state for the diffrax solver for SMEs
+class Y(eqx.Module):
+    rho: Array
+    dYt: Array
+
+
+class MeasurementTerm(dx.ControlTerm):
+    def prod(self, vf: dx.VF, control: dx.Control) -> dx.Y:
+        dW = control
```
Comment (gautierronan) [inline dynamiqs/smesolve/smediffrax.py:23]: Why not just name it `dW` in the arguments? Would be more explicit IMO. This looks like there's additional logic.

File: dynamiqs/smesolve/smediffrax.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+            return Y(rho, dYt)
+
+        lindblad_term = dx.ODETerm(vector_field)
+
+        # === define stochastic term
+        # This contains everything before the "dWt" in the SME:
+        # - measurement backaction term for drho
+        # - simply dWt for dYt
+        def vector_field(t, y, _):  # noqa: ANN001, ANN202
```
Comment (gautierronan) [inline dynamiqs/smesolve/smediffrax.py:76]: Can we name this in a different way? We have two `def vector_field` in the same scope: this doesn't seem very good practice (although I agree it's fine here).

File: dynamiqs/smesolve/smediffrax.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            tmp = Lms_rho + dag(Lms_rho)
+            tr = trace(tmp).real
+
+            # state
+            etas = self.etas[:, None, None]  # (nLm, n, n)
+            tr = tr[:, None, None]  # (nLm, n, n)
+            rho = jnp.sqrt(etas) * (tmp - tr * y.rho)  # (nLm, n, n)
+
+            # signal
+            dYt = jnp.empty(len(self.etas))
```
Comment (gautierronan) [inline dynamiqs/smesolve/smediffrax.py:89]: Should this not be `jnp.zeros` instead? `empty` will just allocate memory without initializing, which might be added to the actual result later on. Maybe I'm missing something.

File: dynamiqs/smesolve/smesolve.py [hunk 5]
```diff
@@ -19,19 +35,19 @@ def smesolve(
...
     options: Options = Options(),  # noqa: B008
-) -> Result:
+) -> SMEResult:
     r"""Solve the diffusive stochastic master equation (SME).
 
     Warning:
-        This function has not been ported to JAX yet. The following documentation is
-        a draft API, copied from the old PyTorch version of the library.
+        This function is still under test and development, it is not as stable as
+        [`dq.sesolve()`][dynamiqs.sesolve] and [`dq.mesolve()`][dynamiqs.mesolve].
```
Comment (gautierronan) [inline dynamiqs/smesolve/smesolve.py:50]:
```suggestion
Please raise any bug you may encounter on github.com/dynamiqs/dynamiqs.
```

File: dynamiqs/smesolve/smesolve.py [hunk 6]
```diff
@@ -63,20 +79,20 @@ def smesolve(
...
 
-    The signals $I_k(t)$ are singular quantities, the solver returns the averaged signals
-    $J_k(t)$ defined for a time interval $[t_0, t_1)$ by:
+    The signals $I_k(t)$ are singular quantities, the solver returns the averaged
+    signals $J_k(t)$ defined for a time interval $[t_0, t_1[$ by:
     $$
-        J_k([t_0, t_1)) = \frac{1}{t_1-t_0}\int_{t_0}^{t_1} I_k(t) \dt
-        = \frac{1}{t_1-t_0}\int_{t_0}^{t_1} \dd y_k(t).
+        J_k([t_0, t_1[) = \frac{1}{t_1-t_0}\int_{t_0}^{t_1} I_k(t)\ \dt
+        = \frac{1}{t_1-t_0}\int_{t_0}^{t_1} \dd Y_k(t).
```
Comment (gautierronan) [inline dynamiqs/smesolve/smesolve.py:95]: I probably missed it, but I thought we were actually returning the `dYt` instead, with the way you wrote it (with the Pytree state). Anyway, just want to make sure :)

File: dynamiqs/smesolve/smesolve.py [hunk 7]
```diff
@@ -91,31 +107,240 @@ def smesolve(
...
+        tutorial for more details
+
+    Warning:
+        Batching on `jump_ops` and `etas is not yet supported, if this is needed don't
+        hesitate to
+        [open an issue on GitHub](https://github.com/dynamiqs/dynamiqs/issues/new).
+
+    Warning:
+        For now the same key is used for every batched simulation over `H` and `rho0`.
+        This will be probably change in the future.
```
Comment (gautierronan) [inline dynamiqs/smesolve/smesolve.py:122]: Isn't this just a matter of splitting the user keys before vmapping?

**Review by gautierronan — CHANGES_REQUESTED** [review]
This looks really great. Very nice work. I have several comments or questions below, but mostly nit. As you stated, there's still some follow-up work to make, but I think it's a great start!

File: dynamiqs/result.py (hunk 1 above)
Comment (pierreguilmin) [reply dynamiqs/result.py:134]: Yes, it's tracked in https://linear.app/dynamiqs/issue/DYN-73/support-generalized-batching.

> I'm thinking a notation like `(nH..., nLs..., nrho0..., ntrajs, ntsave, n, n)` could be more explicit than what we currently have.
Once we've implented the ND batching yes, I like `H: (...H, n, n)` where I read `...H` as "all the other dimensions of H`.

Note that documenting the batching over `jump_ops` is a bit tricky because technically it is `(...L1, ...L2, ..., ...L_nL)`.

File: dynamiqs/smesolve/smediffrax.py (hunk 2 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smediffrax.py:23]: I like using the notations from diffrax API/source code in the overriden function signatures, I think it's neat to understand how our problem maps to the diffrax API. When I read
```python
def prod(self, vf: dx.VF, control: dx.Control) -> dx.Y:
```
I know the `prod` method of the diffrax `ControlTerm` is  reimplemented, while e.g.
```python
def prod(self, rho: dx.VF, dW: dx.Control) -> dx.Y:
```
is more obscure IMO.

In any case, when overriding a base class method it's best practice not to rename the arguments.

File: dynamiqs/smesolve/smediffrax.py (hunk 3 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smediffrax.py:76]: Do you prefer `def vector_field_deterministic()` and `def vector_field_stochastic()`?

File: dynamiqs/smesolve/smediffrax.py (hunk 4 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smediffrax.py:89]: This is correct. In `prod` we just throw this to the bin, and return `Y(rho, dW)` (there is no signal-dependent term in the stochastic part, it's just `dYt = deterministic + dWt`. I think using `jnp.empty` is a good way to express that this is unused, I can maybe put a comment to make it more clear.

File: dynamiqs/smesolve/smesolve.py (hunk 6 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smesolve.py:95]: We indeed accumulate the dYt. The diffrax state at time $t$ is basically

$$\left(\sum_{k=0}^n \mathrm{d}\rho_{k\mathrm{d}t}, \sum_{k=0}^n \mathrm{d}Y_{k\mathrm{d}t}\right)$$

I can add a comment to explain this better too.

But then we do (L104-106 of smediffrax.py):
```python
        # Diffrax integrates the state from t0 to t1. In this case, the state is
        # (rho, dYt). So we recover the signal by simply diffing the resulting array.
        Isave = jnp.diff(integrated_dYt, axis=0)
```
to recover the integrated signal.

File: dynamiqs/smesolve/smediffrax.py (hunk 4 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smediffrax.py:89]: Another option (to explain) would be to do `dYt = jnp.ones(len(self.etas))` and in `prod`:
```
    def prod(self, vf: dx.VF, control: dx.Control) -> dx.Y:
        dW = control
        rho = (vf.rho * dW[:, None, None]).sum(0)  # (n, n)
        dY = vf.dY * dW
        return Y(rho, dY)
```

File: dynamiqs/smesolve/smesolve.py (hunk 6 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smesolve.py:95]: Actually I think the notations are not very rigorous in the code, I will fix this. Can you read this Notion page and tell me if it's clearer? > https://pierreguilmin.notion.site/SME-diffrax-176db6e9e8cc4bdfba4fb30121642625?pvs=4.

Comment (gautierronan) [reply dynamiqs/smesolve/smesolve.py:95]: The notations are good, but I'm still bothered by something. We simulate the continuous SME for which there is an infinitesimal backaction term because we assume a continuous measurement of the quantum system. But we return the integrated signal, which would correspond to a different SME, with a backaction term that only accounts for binned information about the signal.

Am I missing something?

In other words, I feel like there are two equations:

1. the continuous time SME with infinitesimal backaction and continuous signal,
2. the "binned" SME with binned backaction and discrete signals.

I'm not sure about this second equation: does it even exist/make sense?

File: dynamiqs/smesolve/smesolve.py (hunk 7 above)
Comment (gautierronan) [reply dynamiqs/smesolve/smesolve.py:122]: How about both?

By default, the user passes a single key, which we split and distribute among batches (with a different key per batch: seems the most natural to me and what users will expect). And we also allow for passing an array of keys of the same shape as the batch, in which case we don't do the internal splitting and directly pass each key to its corresponding batch.

This second part can come in a separate PR, later on.

File: dynamiqs/smesolve/smesolve.py (hunk 6 above)
Comment (pierreguilmin) [reply dynamiqs/smesolve/smesolve.py:95]: Let me try to clarify. The discrete-time formulation converges (in law) to the continuous SME as the chosen time step $\mathrm{d}t\to0$. This means that you will get the same statistics for the state and signal as the continuous SME if you take a sufficiently small $\mathrm{dt}$. We _could_ return $\mathrm{d}Y_t/\mathrm{d}t$ for each time step $\mathrm{d}t$ we simulate, but (i) this can easily become super costly in memory if there are lots of timestep, (ii) this quantity should be understood with care for adaptive time-stepping and (iii) it can anyway be reconstructed _if needed_ from the Virtual Brownian Tree holding the sampled Wiener process.

I think having the integrated signal in between user-defined times allows for an easier interpretation and rigorous work. But we could have an option to return the full signal too, and all the time steps, that'd be great actually!

**Review by gautierronan — APPROVED** [review]
I forgot what we concluded exactly regarding the API of `smesolve`, but I think this PR looks great as is. I think we should merge and iterate on top with smaller PRs, and we're probably not so far from merging on main too (once we fix the first three bullet points of your to-do list at least, the rest can wait).
